File: XGBoost/XGBoost.R
# XGBoost

#Read in data
dataset <- read.csv("Churn_Modelling.csv")
dataset <- dataset[4:14]

#Encode the categorical variables as factors (and converting to numeric for the NN package)

dataset$Geography = as.numeric(factor(dataset$Geography, 
                                      levels = c("Germany", "France", "Spain"),
                                      labels = c(1,2,3)))

dataset$Gender = as.numeric(factor(dataset$Gender, 
                                   levels = c("Female", "Male"),
                                   labels = c(1,2)))

#Partition the data
library(caTools)
set.seed(123)

split = sample.split(dataset$Exited, SplitRatio = 0.8)
training_set = subset(dataset, split==TRUE)
test_set = subset(dataset, split==FALSE)

#Fitting XGBoost to the Training Set
library(xgboost)
classifier = xgboost(data = as.matrix(training_set[-11]), label=training_set$Exited, nrounds = 10)

#Applying k-Fold Cross Validation
library(caret)
folds = createFolds(training_set$Exited, k=10)
cv = lapply(folds, function(x) {
  training_fold = training_set[-x, ] #remove the current fold x
  test_fold = training_set[x,] #only the current fold x
  classifier = xgboost(data = as.matrix(training_set[-11]), label=training_set$Exited, nrounds = 10)
  y_pred <- predict(classifier, newdata = as.matrix(test_fold[-11]))
  y_pred = (y_pred > 0.5)
  cm <- table(test_fold[,11],y_pred)
  accuracy = (cm[1,1] + cm[2,2]) / (cm[1,1] + cm[2,2] + cm[1,2] + cm[2,1])
  return(accuracy)
})

accuracy = mean(as.numeric(cv))

#Predict the test set result
prob_pred <- h2o.predict(classifier, newdata = as.h2o(test_set[-11]))
y_pred = ifelse(prob_pred > 0.5, 1, 0)
y_pred = as.vector(y_pred)

#Making the confusion matrix
cm <- table(test_set[,11],y_pred)

File: XGBoost/XGBoost.py
#XG Boost

#Taking code from ANN codeset

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Step 1: Data Preprocessing

#Importing the dataset
dataset = pd.read_csv("Churn_Modelling.csv")
X = dataset.iloc[:,3:13].values
y = dataset.iloc[:,13].values

# Encoding categorical data
# Encoding the Independent Variable
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
#label the country feature
labelencoder_X_1 = LabelEncoder()
X[:, 1] = labelencoder_X_1.fit_transform(X[:, 1])
#label the gender feature
labelencoder_X_2 = LabelEncoder()
X[:, 2] = labelencoder_X_2.fit_transform(X[:, 2])
#perform one-hot encoding
onehotencoder = OneHotEncoder(categorical_features = [1])
X = onehotencoder.fit_transform(X).toarray()
#Avoid the dummy variable trap
X = X[:,1:]

#Splitting the dataset into training and test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test  = train_test_split(X,y,test_size=0.25, random_state=0)

#Fitting XGBoost to the Training Set
from xgboost import XGBClassifier
classifier = XGBClassifier(learning_rate=0.2, gamma=0.2)
classifier.fit(X_train, y_train)

#Predict the Test set results
y_pred = classifier.predict(X_test)

#Creating a confusion matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test,y_pred)

#Applying k-Fold Cross Validation 
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = classifier, X = X_train, y=y_train, cv = 10)
#we take the mean and standard deviation of these 10 validations
accuracies.mean()
accuracies.std()

#Applying Grid Search to find the best model and the best parameters
from sklearn.model_selection import GridSearchCV
#in the below dictionary, we create two entries which contain the parameters as keys and the parameter settings that we want to test out as the values
parameters = [{'max_depth' : [3], 'learning_rate' : [0.2], 'gamma' : [0.2], 'n_estimators' : [10, 100, 200, 1000]},
              ]
grid_search = GridSearchCV(estimator = classifier, param_grid = parameters, scoring = 'accuracy', cv = 10, n_jobs = -1)

grid_search = grid_search.fit(X_train, y_train)

#gives us the best accuracy, remember this is using k-fold cross validation
best_accuracy = grid_search.best_score_

#gives us a list of the best parameters selected
best_params = grid_search.best_params_